### PR TITLE
Use Failure instead of exception in ProblemEvents

### PR DESCRIPTION
**Pull request by reinsch82**
Fixes 
  - #27755

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

File: platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r85/ProblemsServiceModelBuilderCrossVersionTest.groovy [hunk 1]
```diff
@@ -82,8 +83,8 @@ class ProblemsServiceModelBuilderCrossVersionTest extends ToolingApiSpecificatio
         """
     }
 
-    @TargetGradleVersion(">=8.6")
-    @ToolingApiVersion(">=8.6")
+    @TargetGradleVersion(">=8.7")
```
Comment (donat) [inline platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r85/ProblemsServiceModelBuilderCrossVersionTest.groovy:86]: ❌ he test coverage should not be changed for existing releases. Instead, an upper bound should be defined (`@TargetGradleVersion(">=8.6 <8.7")`) and the new behaviour for 8.7 should be defined in a separate test class in the `r87` package.

**Review by donat — CHANGES_REQUESTED** [review]

**Comment by reinsch82** [comment]
@bot-gradle test this

File: platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r86/ProblemsServiceModelBuilderCrossVersionTest.groovy [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        ProblemProgressListener listener = new ProblemProgressListener()
+
+        when:
+        withConnection {
+            it.model(CustomModel)
+                .setJavaHome(jdk.javaHome)
+                .addProgressListener(listener)
+                .get()
+        }
+        def problems = listener.problems.collect{ it as ProblemDescriptor}
```
Comment (donat) [inline platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r86/ProblemsServiceModelBuilderCrossVersionTest.groovy:50]: 💅 

```suggestion
        def problems = listener.problems.collect{ it as ProblemDescriptor }
```

File: platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/ProblemDescriptor.java [hunk 3]
```diff
@@ -88,11 +88,10 @@ public interface ProblemDescriptor extends BaseProblemDescriptor {
     List<Solution> getSolutions();
 
     /**
-     * Returns the exception associated with this problem.
+     * Returns the failure associated with this problem.
```
Comment (donat) [inline platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/ProblemDescriptor.java:91]: ❓ Should we document what happens in the TAPI 8.7 -> Gradle 8.6 scenario (presumably `getFailure()` will always return `null`).  Do we have a test to verify that?

File: platforms/ide/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/parameters/BuildProgressListenerAdapter.java [hunk 4]
```diff
@@ -793,8 +794,15 @@ private static List<ProblemDescriptor> toProblemDescriptors(Collection<InternalP
...
     }
 
-    private static ExceptionContainer toExceptionContainer(@Nullable RuntimeException exception) {
-        return new DefaultExceptionContainer(exception);
+    private static @Nullable FailureContainer toExceptionContainer(@Nullable InternalBasicProblemDetails exception) {
+        if (!(exception instanceof InternalBasicProblemDetails2)) {
+            return null;
+        }
+        InternalFailure failure = ((InternalBasicProblemDetails2) exception).getFailure();
+        if(failure == null){
```
Comment (donat) [inline platforms/ide/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/parameters/BuildProgressListenerAdapter.java:802]:
```suggestion
        if (failure == null) {
```

File: platforms/ide/tooling-api/src/main/java/org/gradle/tooling/internal/protocol/problem/InternalBasicProblemDetails2.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+ * limitations under the License.
+ */
+
+package org.gradle.tooling.internal.protocol.problem;
+
+import org.gradle.tooling.internal.protocol.InternalFailure;
+
+import javax.annotation.Nullable;
+
+public interface InternalBasicProblemDetails2 extends InternalBasicProblemDetails {
```
Comment (donat) [inline platforms/ide/tooling-api/src/main/java/org/gradle/tooling/internal/protocol/problem/InternalBasicProblemDetails2.java:23]: ❓ The convention in the protocol package is to use the `VersionX` suffix for classes, i.e. this type probably should be named `InternalBasicProblemDetailsVersion2`.

File: subprojects/build-events/src/main/java/org/gradle/internal/build/event/types/DefaultProblemDetails.java [hunk 6]
```diff
@@ -41,7 +42,7 @@ public class DefaultProblemDetails implements InternalBasicProblemDetails, Seria
     private final InternalDocumentationLink documentationLink;
     private final List<InternalSolution> solutions;
     private final InternalAdditionalData additionalData;
-    private final RuntimeException exception;
+    private final InternalFailure exception;
```
Comment (donat) [inline subprojects/build-events/src/main/java/org/gradle/internal/build/event/types/DefaultProblemDetails.java:45]:
```suggestion
    private final InternalFailure failure;
```

File: platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r86/ProblemProgressEventCrossVersionTest.groovy [hunk 7]
```diff
@@ -183,7 +184,8 @@ class ProblemProgressEventCrossVersionTest extends ToolingApiSpecification {
         return listener.problems.collect { (ProblemDescriptor) it }
     }
 
-    @ToolingApiVersion(">=8.6")
+    @ToolingApiVersion(">=8.7")
+    @TargetGradleVersion(">=8.7")
```
Comment (donat) [inline platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r86/ProblemProgressEventCrossVersionTest.groovy:188]: ❌ These tests are unrelated to the the intent of this PR. Why changing the version ranges?

File: platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r86/ProblemProgressEventCrossVersionTest.groovy [hunk 8]
```diff
@@ -124,7 +124,8 @@ class ProblemProgressEventCrossVersionTest extends ToolingApiSpecification {
         }
     }
 
-    @ToolingApiVersion(">=8.6")
+    @ToolingApiVersion(">=8.7")
+    @TargetGradleVersion(">=8.7")
```
Comment (donat) [inline platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r86/ProblemProgressEventCrossVersionTest.groovy:128]: ❌  This test probably should live in the `r87` package.

**Review by donat — CHANGES_REQUESTED** [review]

File: platforms/core-runtime/launcher/src/test/groovy/org/gradle/launcher/daemon/protocol/DaemonMessageSerializerTest.groovy [hunk 9]
```diff
@@ -28,11 +28,11 @@ import org.gradle.internal.logging.events.UserInputResumeEvent
 import org.gradle.internal.serialize.PlaceholderException
 import org.gradle.internal.serialize.Serializer
 import org.gradle.internal.serialize.SerializerSpec
-import org.gradle.tooling.internal.provider.action.BuildActionSerializer
-import org.gradle.tooling.internal.provider.action.ExecuteBuildAction
 import org.gradle.launcher.daemon.diagnostics.DaemonDiagnostics
 import org.gradle.launcher.exec.BuildActionResult
 import org.gradle.launcher.exec.DefaultBuildActionParameters
+import org.gradle.tooling.internal.provider.action.BuildActionSerializer
+import org.gradle.tooling.internal.provider.action.ExecuteBuildAction
```
Comment (donat) [inline platforms/core-runtime/launcher/src/test/groovy/org/gradle/launcher/daemon/protocol/DaemonMessageSerializerTest.groovy:35]: 💅 The file is practically unchanged;  It's fine to leave it in, but in the future I'd remove them to minimize PR footprint.

File: platforms/ide/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/ProblemsProgressEventConsumer.java [hunk 10]
```diff
@@ -103,11 +105,20 @@ private DefaultProblemEvent createProblemEvent(OperationIdentifier buildOperatio
...
-                toInternalAdditionalData(problem.getAdditionalData())
+                toInternalAdditionalData(problem.getAdditionalData()),
+                toInternalFailure(problem.getException())
             )
         );
     }
 
+    @Nullable
+    private static InternalFailure toInternalFailure(@Nullable RuntimeException ex) {
+        if(ex == null) {
```
Comment (donat) [inline platforms/ide/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/ProblemsProgressEventConsumer.java:116]: 💅 

```suggestion
        if (ex == null) {
```

File: platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r85/ProblemProgressEventCrossVersionTest.groovy [hunk 11]
```diff
@@ -17,34 +17,50 @@
...
 class ProblemProgressEventCrossVersionTest extends ToolingApiSpecification {
 
-    class MyProgressListener implements ProgressListener {
-        List<ProblemDescriptor> allProblems = []
-
-        @Override
-        void statusChanged(ProgressEvent event) {
-            if (event instanceof ProblemEvent) {
-                this.allProblems.addAll(event.getDescriptor())
+    def "test failure context"() {
```
Comment (donat) [inline platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r85/ProblemProgressEventCrossVersionTest.groovy:31]: ❌ The two tests in this class seems identical in terms of covered scenarios as well as the method bodies. The only difference is this bit:
```
problems.every {
            it == [:]
        }
```

Can we merge the the two test methods?

File: platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r85/ProblemsServiceModelBuilderCrossVersionTest.groovy [hunk 12]
```diff
@@ -85,78 +87,21 @@ class ProblemsServiceModelBuilderCrossVersionTest extends ToolingApiSpecificatio
...
         }
-
         then:
         listener.problems.size() == 1
-        listener.problems[0].additionalData.asMap == [
-            'keyToString': 'value'
-        ]
     }
 
     @TargetGradleVersion(">=8.6")
```
Comment (donat) [inline platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r85/ProblemsServiceModelBuilderCrossVersionTest.groovy:107]: ❌ This doesn't seem right. I'm getting lost in the test reshuffling, but I assume we should have something like 

```
    @ToolingApiVersion(">=8.6")
    @TargetGradleVersion(">=8.5 <8.7")
    def "getFailure always null in older versions"() { ... }
```

in the `r86` package.

File: platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r86/ProblemsServiceModelBuilderCrossVersionTest.groovy [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+import org.gradle.tooling.events.problems.ProblemDescriptor
+import org.junit.Assume
+
+import static org.gradle.integtests.fixtures.AvailableJavaHomes.getJdk17
+import static org.gradle.integtests.fixtures.AvailableJavaHomes.getJdk21
+import static org.gradle.integtests.fixtures.AvailableJavaHomes.getJdk8
+import static org.gradle.integtests.tooling.r85.ProblemsServiceModelBuilderCrossVersionTest.getBuildScriptSampleContent
+
+@TargetGradleVersion(">=8.6")
+@ToolingApiVersion(">=8.6")
```
Comment (donat) [inline platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r86/ProblemsServiceModelBuilderCrossVersionTest.groovy:33]: This annotation is practically unused.

File: platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r87/ProblemProgressEventCrossVersionTest.groovy [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
+
+    def withReportProblemTask(@GroovyBuildScriptLanguage String taskActionMethodBody) {
+        buildFile getProblemReportTaskString(taskActionMethodBody)
+        // TODO using the following code breaks the test, but it should be possible to use it
+        //  buildFile getProblemReportingScript(taskActionMethodBody)
+        //  issue https://github.com/gradle/gradle/issues/27484
+    }
+
+    @TargetGradleVersion("=8.5")
+    def "test failure context"() {
```
Comment (donat) [inline platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r87/ProblemProgressEventCrossVersionTest.groovy:48]: The name of the method is cryptic. What are we testing here? I know, but so should the next person stumbling upon this class.

```suggestion
    def "test failure context"() {
```

File: platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r87/ProblemProgressEventCrossVersionTest.groovy [hunk 15]
```diff
@@ -0,0 +1,279 @@
...
+                .addArguments("--info")
+                .run()
+        }
+
+        then:
+        thrown(BuildException)
+        listener.problems.size() == 2
+    }
+
+    def "Problems expose details via Tooling API events"() {
```
Comment (donat) [inline platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r87/ProblemProgressEventCrossVersionTest.groovy:77]: This and the tests below already worked with TAPI 8.6 -> Gradle 8.6 and so they belong to the `r86` package, without the failure assertion of course. The tests should be moved back to the `r86` package and then have a simplified variant here, asserting the exception field.

File: platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r87/ProblemsServiceModelBuilderCrossVersionTest.groovy [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+        then:
+        thrown(BuildException)
+        def problems = listener.problems.collect { it as ProblemDescriptor }
+        problems.size() == 1
+        problems[0].label.label == "Could not compile build file '$buildFile.absolutePath'."
+        problems[0].category.category == 'compilation'
+        problems[0].failure.failure.message == "Could not compile build file '$buildFile.absolutePath'."
+    }
+
+    def "Can use problems service in model builder"() {
```
Comment (donat) [inline platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r87/ProblemsServiceModelBuilderCrossVersionTest.groovy:63]: Probably same as above, in  platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r87/ProblemProgressEventCrossVersionTest.groovy

File: platforms/ide/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/DefaultFileComparisonTestAssertionFailure.java [hunk 17]
```diff
@@ -29,7 +29,15 @@ public class DefaultFileComparisonTestAssertionFailure extends DefaultTestAssert
     private final byte[] expectedContent;
     private final byte[] actualContent;
 
-    public DefaultFileComparisonTestAssertionFailure(String message, String description, String expected, String actual, List<? extends Failure> causes, String className, String stacktrace, byte[] expectedContent, byte[] actualContent) {
+    public DefaultFileComparisonTestAssertionFailure(@Nullable String message,
```
Comment (donat) [inline platforms/ide/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/DefaultFileComparisonTestAssertionFailure.java:32]: ❌ This change is unrelated to the PR and actually breaks the formatting also used in `DefaultTestAssertionFailure`, `DefaultTestFrameworkFailure`.

**Review by donat — CHANGES_REQUESTED** [review]
The production code seems fine to me, but I have some comments, mainly on the test coverage.

File: platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r87/ProblemProgressEventCrossVersionTest.groovy (hunk 15 above)
Comment (reinsch82) [reply platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r87/ProblemProgressEventCrossVersionTest.groovy:77]: The newer one also has to check for `failure` to my understanding.

**Comment by reinsch82** [comment]
@bot-gradle test ACT

**Review by donat — APPROVED** [review]

**Comment by reinsch82** [comment]
@bot-gradle merge
